daemon/capabilities/capabilities_versions.go

```diff
@@ -0,0 +1,9 @@
+package capabilities
+
+// CapabilitiesV1 must implement Capabilities
+var _ VersionedCapabilities = &capabilitiesV1{}
+
+type capabilitiesV1 struct {
+	CapabilitiesBase
+	RegistryClientAuth bool `json:"registry-client-auth"`
```

Comment:
Wondering if this is _too strict_. I was considering that the type we use (for the API response) could use a `map[string]XXX` (`bool` or  `any` ?), but we can define a list of capabilities (which can be versioned);


```go
type CapabilitiesBase struct {
	// CapabilitiesVersion is a natural number representing the version
	// of the capabilities API this response is for.
	//
	// This is part of the Capabilities version negotiation flow:
	// Since the API contract states that any client requesting
	// capabilities version N MUST also accept any capabilities
	// version M such that 1 < M < N, the daemon can pick a version
	// M = min(DAEMON_CAPABILITIES_VERSION, N), and send back a
	// response for that version.
	CapabilitiesVersion Version `json:"_v"`

	Capabilities map[string]any `json:"capabilities"`
}
```

This would preserve the option for 3-rd party capabilities (in the Docker Desktop, Mirantis MCR) to be added in their API proxy.

Not exactly sure what the best way is to document those capabilities in that case; perhaps something similar to how we document the runtime properties, where we document them as "well-known" ones, but in this case it may be slightly more strict, so perhaps not "well known" but a list of capabilities that may be returned 🤔 https://github.com/moby/moby/blob/01b98afeb65b4a5040d96958326fa04f5ab12c72/api/swagger.yaml#L6141-L6161

Reply:
IMO, we can't have both:
- stable, solid version negotiation handling with backwards/forwards guarantees, and
- support for arbitrary values to be used by third parties

This also makes the response parsing logic on the client side significantly more complex. I'd rather 3rd party clients (which already must be aware of custom features added by their provider) parse the response themselves than explicitly support it.

If we really want to, I'd rather take an option on the client only to parse additional fields and return a `map[string]string`, or even to allow the client to pass in an `any` to be used for deserializing rather than change the types, which should be the canonical representation of what we actually support.

Reply:
@corhere thoughts?

Reply:
I'm thinking we should reserve the capabilities map for upstream Moby and afford other ways to augment the ping response with arbitrary vendor extensions. That cleanly sidesteps any possibility of naming collisions between third-party capabilities and future Moby capabilities or third-party capabilities from another vendor. To that end, what if we versioned the ping response as a whole? In practical terms that would just be changing the query parameter name and flattening the JSON document by one level.

E.g.: `GET /ping?v=1` could return something along the lines of
```go
type PingBody struct {
        Version      uint                        `json:"_v"`
        Capabilities CapabilitiesV1              `json:"capabilities"`
        Extensions   map[string]json.RawMessage  `json:"extensions,omitempty"`
}
```

We could give vendors a sandbox to pass arbitrary data in the response which can easily be parsed out of the fixed structure and passed through the Engine API client library, with some guidelines (like naming their extension after a reversed DNS name they own) to avoid name collisions between vendors. We can therefore reserve all properties of the root JSON object for future non-breaking upstream additions to the ping response.

```json
{
  "_v": 1,
  "capabilities": {
    "registry-client-auth": true
  },
  "extensions": {
    "com.docker.desktop": {
      "capabilities-with-a-different-schema-because-why-not": ["foo", "bar", "!baz"],
      "hello": "world"
    }
  }
}
```

With the vendor extensions out of the way, the capabilities Go type can be defined as a struct rather than a map without issue as it does not need to be generic or extensible. The capability fields can be any type, and they can be strongly typed for each capability! It does not matter if the daemon sends a capability that the client doesn't know about; the client will silently ignore it.